feat(ingest): datamodel to ingest organisation role metadata for a dataset

sheeru opened the pull request:
The aim of this feature is to enable users to view the required roles (Roles/Policies defined in Access Management System of an organisation) for accessing a dataset in that organisation. Also, should be able to request for the appropriate roles from the Datahub frontend to the organisation via urls.

Will attach details RFC document


## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

jjoyce0510 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -1397,6 +1404,56 @@ type Dataset implements EntityWithRelationships & Entity & BrowsableEntity {
+    """
+    relationships(input: RelationshipsInput!): EntityRelationshipsResult
+
+    id: String!
```
Comments please1

sheeru replied:
Done

jjoyce0510 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -1397,6 +1404,56 @@ type Dataset implements EntityWithRelationships & Entity & BrowsableEntity {
+
+    properties: ExternalRoleProperties!
+
+    provisionedUsers: [CorpUser!]
```
Let's put this inside a wrapper object... Called "users"

jjoyce0510 replied:
Or better yet - Actors

sheeru replied:
Done

jjoyce0510 commented on metadata-models/src/main/pegasus/com/linkedin/externalrolemetadata/ExternalRoleProvisionedUser.pdl:
```diff
@@ -0,0 +1,20 @@
+ * Provisioned users of a role
+ */
+@Aspect = {
+  "name": "externalRoleProvisionedUser"
```
What will happen when the requirement to add groups comes along? How about other nested roles? 

I'd rename this aspect to "actors", so that we can support users, groups, and nested roles

sheeru replied:
Have renamed to actors and now we have the ability to add groups under actors when we require

jjoyce0510 commented:
Hi @sheeru ! 

Thanks for the PR.

A few things I want us to consider, in addition to those comments I've already left:

- Have we considered a "live lookup" approach? Wherein DataHub calls your service to determine 

      1. whether a particular user has access to a particular dataset 
      2. request access for a particular user to a particular dataset (or role) 

If this does not work, why not? The reason I ask is that the changes required to support this in DataHub are a bit more constrained. 

- How can we limit the blast radius of these changes? A few things I'd request if we do go with this approach:

     1.  Let's rename externalRoleMetadata to just "role". A role will be an external role, by default. dataHubRole is already an entity with a name that indicates it's usage within the DataHub system. 
     2. Let's avoid defining an enum for PrivilegeType and simply use a string to denote the privileges associated with the role. This is because each system we want to integrate with in the future may have a different "native privilege type" concept. For example, in some SQL systems you do not have READ, WRITE, etc, but instead have "SELECT" or "INSERT" or "DELETE". 
     3. Is there any chance a _group_ or another _role_ may be associated with a Role? If yes, I'd want to have some plan to support group as something a role can point to. 
     4. Let's rename the externalRolesMetadata aspect of dataset entity to simple "access". This can grow to include additional access related information relevant for the dataset over time. 

Let me know if you have questions!

Cheers
John

sheeru commented:
@jjoyce0510 Have addressed the comments and pls let me know if anything more is required.
Regarding the 2 points which u have highlighted,

```
1. whether a particular user has access to a particular dataset 
2. request access for a particular user to a particular dataset (or role) 
```

Since we are accessing the dataset from UI, we will know who has logged in. We will compare if the current logged in user is present as part of provisioned users and we can determine if the user has access to a role and dataset. Attached in RFC for more details

jjoyce0510 commented:
Thanks for attaching the RFC, I've taken a look at it.

I am very curious to understand _why_ the listing of roles is so important here, as opposed to showing simpler things: does the user have access or not, and how can the user request access? 

I'm imagining a flow where DataHub can simply redirect to an external URL where the role can be finally selected from a list of roles which have access. 

We'd likely support a pluggable backend component (in the GraphQL server) that accepts 2 inputs: 

- user urn
- dataset urn

and returns a simple boolean which represents whether the user urn has access to the dataset urn. We'd also allow for plugging in a "request access url" which would be a custom url that is redirected to when the user clicks to request access. 

If we can have an approach like this, we have 2 major benefits:

1. **Simplicity, Separation of Concerns**: We do not need to manage synchronizing the external Roles and their users into DataHub and keep this mapping up to date. Instead, that information stays in the external system and is simply queried in real time. This reduces the overhead your team will need to spend on keeping information up to date in DataHub and reduce the possibility of difficult synchronization bugs. 

2. **Extensibility, Generalization**: It is not clear to me that the proposed approach will generalize well across other organizations, who may not have the same requirements around requesting access for a specific role. By using a live-lookup approach, we can completely separate DataHub from the external system and provide an interface which can be easily extended to support new use cases as they arise. 

So given these benefits, please help me to understand why the proposed approach will be better in the long term and applicable across multiple organizations who may use it...  

Also, few other responses on the thread:

> Also, PrivilegeType is defined as enum because, UI should be able to group the access types and show the roles under 
> that group. Hence enum is preferred to handle the list of access types.

We can easily group inside the UI even if we do not use enum. Please change to be a string for now. We can always add back an enum later once the Domain becomes more clear. My goal with this PR is to make it as easy as possible to extend beyond your use cases! 

Cheers
John

jjoyce0510 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -1399,6 +1409,84 @@ type Dataset implements EntityWithRelationships & Entity & BrowsableEntity {
+}
+
+type Actor {
+    provisionedUsers: [CorpUser!]
```
comments please!

sheeru replied:
Done

jjoyce0510 commented on li-utils/src/main/javaPegasus/com/linkedin/common/urn/RoleUrn.java:
```diff
@@ -0,0 +1,68 @@
+import java.net.URISyntaxException;
+
+
+public final class RoleUrn extends Urn {
```
This class is not necessary, feel free to remove!

sheeru replied:
Done

jjoyce0510 commented on metadata-models/src/main/pegasus/com/linkedin/common/Access.pdl:
```diff
@@ -0,0 +1,13 @@
+}
+record Access {
+
+  roles: array[RoleAssociation]
```
comments please!

sheeru replied:
Done

jjoyce0510 commented on metadata-models/src/main/pegasus/com/linkedin/common/Access.pdl:
```diff
@@ -0,0 +1,13 @@
+}
+record Access {
+
+  roles: array[RoleAssociation]
```
let's also make this field optional

sheeru replied:
Done

jjoyce0510 commented on metadata-models/src/main/pegasus/com/linkedin/common/RoleAssociation.pdl:
```diff
@@ -0,0 +1,17 @@
+    "name": "AssociatedWith",
+    "entityTypes": [ "role" ]
+  }
+  urn: RoleUrn
```
This can just be "Urn", we recommend against using strongly-typed urns now. (they are legacy)

sheeru replied:
Done

jjoyce0510 commented on metadata-models/src/main/pegasus/com/linkedin/role/RoleProperties.pdl:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * Link to access external access management
+   */
+  requesturl: optional string
```
Camel case: requestUrl

sheeru replied:
Done

jjoyce0510 commented on metadata-models/src/main/pegasus/com/linkedin/role/RoleProperties.pdl:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * Can be READ, ADMIN, WRITE
+   */
+  type: string
```
Thanks for making this a string

jjoyce0510 commented on metadata-models/src/main/pegasus/com/linkedin/role/Actors.pdl:
```diff
@@ -0,0 +1,17 @@
+  /**
+   * List of provisioned users of a role
+   */
+  provisionedUsers: array[RoleProvisionedUser]
```
let's just name this more simply: users

sheeru replied:
Done

jjoyce0510 reviewed (commented):
For the current approach, I left some naming comments. Let's consider the above comment about the requirements and pros / cons first ^

jjoyce0510 commented on datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/types/dataset/DatasetType.java:
```diff
@@ -85,7 +85,8 @@ public class DatasetType implements SearchableEntityType<Dataset, String>, Brows
         DATA_PRODUCTS_ASPECT_NAME,
-        BROWSE_PATHS_V2_ASPECT_NAME
+        BROWSE_PATHS_V2_ASPECT_NAME,
+            ACCESS_DATASET_ASPECT_NAME
```
indenting is inconsistent it seems!

sheeru replied:
Done

jjoyce0510 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -1399,6 +1409,84 @@ type Dataset implements EntityWithRelationships & Entity & BrowsableEntity {
+    roles: [RoleAssociation!]
+}
+
+type Role implements Entity{
```
nit: Entity { 

add a space

sheeru replied:
Done

jjoyce0510 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -1399,6 +1409,84 @@ type Dataset implements EntityWithRelationships & Entity & BrowsableEntity {
+}
+
+type Actor {
+    provisionedUsers: [CorpUser!]
```
let's name this users

sheeru replied:
Done

jjoyce0510 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -1399,6 +1409,84 @@ type Dataset implements EntityWithRelationships & Entity & BrowsableEntity {
+}
+
+type Actor {
+    provisionedUsers: [CorpUser!]
```
Question: Why are we have a RoleProvisionedUser object in the GMS Backend layer but we have a list of resolved users here.

My expectation is to see:

```
users: [RoleUser!]! 


...

type RoleUser {
     // The user attached to the role
      user: CorpUser!
}
```

here.

sheeru replied:
Got it. Done

jjoyce0510 commented on datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/resolvers/EntityTypeMapper.java:
```diff
@@ -16,6 +16,7 @@ public class EntityTypeMapper {
   static final Map<EntityType, String> ENTITY_TYPE_TO_NAME =
       ImmutableMap.<EntityType, String>builder()
           .put(EntityType.DATASET, "dataset")
+          .put(EntityType.ROLE, "externalRoleMetadata")
```
This should be "role"

sheeru replied:
Yup. Done

sheeru commented:
@jjoyce0510  
Also, every table in our organisation has more than one roles available to access underlying table - Basic role, role that gives access to PII data, Role that gives access to PCI data, Admin Role, operations role. The user will have to request the one based on his/her requirement. Since there is no 1-1 mapping between role/table - the feature was designed that way.

The flow was specifically designed based on the user needs so they can raise for appropriate role from Datahub itself to save time. The search doesnt end at Datahub, it just begins and user go to IAM role or Notebooks depending on their access. Please note that this is one of the step as part of Data Discovery which we are trying to solve as part of Datahub.

sheeru commented on metadata-models/src/main/pegasus/com/linkedin/role/RoleProvisionedUser.pdl:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Provisioned users of a role
+ */
+@Aspect = {
```
Done

sheeru commented on metadata-models/src/main/pegasus/com/linkedin/role/RoleProvisionedUser.pdl:
```diff
@@ -0,0 +1,20 @@
+      "name": "Has",
+      "entityTypes": [ "corpuser" ]
+    }
+    provisionedUser: Urn
```
Done

sheeru commented on metadata-models/src/main/pegasus/com/linkedin/role/RoleProvisionedUser.pdl:
```diff
@@ -0,0 +1,20 @@
+@Aspect = {
+  "name": "roleProvisionedUser"
+}
+record RoleProvisionedUser {
```
Done

jjoyce0510 commented on datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GmsGraphQLEngine.java:
```diff
@@ -726,6 +751,7 @@ private void configureQueryResolvers(final RuntimeWiring.Builder builder) {
             .dataFetcher("browse", new BrowseResolver(browsableTypes))
             .dataFetcher("browsePaths", new BrowsePathsResolver(browsableTypes))
             .dataFetcher("dataset", getResolver(datasetType))
+            .dataFetcher("role", getResolver(roleType))
```
I have one final comment. Even though internally the model is Role, let's call the external resolver "externalRole". I didn't realize that we already have another Query called "listRoles", so at least at GraphQL we will need this distinction, even if the underlying PDL models do not have it.

jjoyce0510 replied:
(see line 800)

jjoyce0510 commented on datahub-graphql-core/src/main/resources/entity.graphql:
```diff
@@ -1399,6 +1409,94 @@ type Dataset implements EntityWithRelationships & Entity & BrowsableEntity {
+    roles: [RoleAssociation!]
+}
+
+type Role implements Entity {
```
As discussed offline, ExternalRole is indeed safer

jjoyce0510 commented on li-utils/src/main/java/com/linkedin/metadata/Constants.java:
```diff
@@ -228,6 +228,14 @@ public class Constants {
   public static final String DOMAINS_ASPECT_NAME = "domains";
+
+  // ExternalRoleMetadata
+  public static final String ROLE_ENTITY_NAME = "role";
```
As described offline, externalRole

jjoyce0510 commented on metadata-models/src/main/pegasus/com/linkedin/common/Access.pdl:
```diff
@@ -0,0 +1,15 @@
+  /**
+   * List of Roles which needs to be associated
+   */
+  roles: optional array[RoleAssociation]
```
externalRoles - as described offline

jjoyce0510 commented:
Once this is green, we will merge. I will take over and fix some of the naming issues that now exist.

jjoyce0510 commented:
(Approved for run - hoping to merge this tomorrow morning PST)

jjoyce0510 reviewed (approved):
LGTM! Thanks for the hard work here. 

Cheers
John

lix-mms commented:
Although this change is great and solves important problems, we see some limitations:
- For organisations like ours, with auditing being very significant, we need to be able to view histories of changes to individual role assignment. It is not convenient for managing the life-cycle of individual role assignments based on this PR.
- The assignment of more than one roles (in our case instead of roles we use platform principals like groups, service accounts, etc) might be batched into the same access request. Based this PR the modelling of such batches would be quite difficult.

Our organisation is working on a data model with more potential feature support, and fortunately it does not seem to conflict with the above change so far 🍀. If anyone wishes we are keen to align with more details 😊
